### PR TITLE
Feature Optional prophet

### PR DESCRIPTION
**Pull request by mauicv**
# Todo:
- [x] Tox env tests
- [ ] There is at least one fix in #553 that are needed here in order for tox tests to pass :face_exhaling: namely: tox env needs pytest-randomly dependency
___

## What is this

This work branches off [`feature/dep-management`](https://github.com/SeldonIO/alibi-detect/pull/501) and will make `prophet` optional dependency of `alibi-detect`.

___

### What this PR doesn't include:

1. The install options in the docs. I think it makes sense to do this in a later PR on the feature branch for all the optional deps all at once, once this branch has been merged in.
2. The install options in the `README.md` for same reason as 1.
3. The config-driven work in the [release branch](https://github.com/SeldonIO/alibi-detect/tree/release/0.10.0). This will be done in a separate PR in order to manage Release/Master branch complexity issues.
4. Make Licenses updates. This will be done in a PR on the feature branch.

___

### CODE REVIEW

File: alibi_detect/cd/pytorch/context_aware.py
Comment (ascillitoe) [inline alibi_detect/cd/pytorch/context_aware.py:268]: **Note**: @jklaise if when you come to review you're wondering why the error is only apparent now, I suspect it is because `numpy>=1.23.0` is now installable due to the change in core deps. See https://github.com/SeldonIO/alibi-detect/pull/539#issuecomment-1167205579.

File: alibi_detect/cd/pytorch/mmd.py
Comment (ascillitoe) [inline alibi_detect/cd/pytorch/mmd.py:121]: **Note**: This error is discussed here: https://github.com/SeldonIO/alibi-detect/issues/540

File: alibi_detect/tests/test_dep_management.py
Comment (ascillitoe) [inline alibi_detect/tests/test_dep_management.py:104]: Why do we no longer do the `models.pytorch` submodule tests?

File: setup.cfg
Comment (ascillitoe) [inline setup.cfg:33]: Why are `tensorflow` and `torch` commented out now?

**Review by ascillitoe — COMMENTED** [review]
Generally LGTM, but I just have a few queries about the tests.

File: setup.cfg
Comment (ascillitoe) [reply setup.cfg:33]: Perhaps answering my own question, is this because #538 is not merged into `feature/dep-management` yet? Perhaps I should hold off on reviewing the sub-PR's until that is merged?

Comment (mauicv) [reply setup.cfg:33]: No, basically I added these all in the dep-management branch and then branched off of that. This meant the tox tests ran for all the environments in this branch which was annoying because I only really wanted to test the prophet dep bucket. Basically, as a quick fix, I just commented them out or also deleted irrelevant tests above. I'm aware that this is going to mean being very careful when merging. I think in retrospect I should have done a series of branches one of each other so: 

`dep-management -> optional-backends -> optional prophet -> optional numba`

rather than trying to isolate all the common code in the feature branch. In the end, I found myself constantly having to go back and fix something in the feature branch and then re-merging it into the work branches... The core mistake was just having code that's relevant to specific optional dependencies defined in the `dep-management` branch. 

I'm happy to update this if you think it would significantly reduce the risk of introducing errors at the point I merge the working branches into the feature branch. My current plan was to be very careful...

Comment (ascillitoe) [reply setup.cfg:33]: I am happy with you being very careful 😅 

Assuming we still have a final opportunity to check via the PR into `master`?

Comment (mauicv) [reply setup.cfg:33]: haha, yeah of course

File: alibi_detect/tests/test_dep_management.py
Comment (mauicv) [reply alibi_detect/tests/test_dep_management.py]: We will, see [this](https://github.com/SeldonIO/alibi-detect/pull/541#discussion_r913659850)

File: alibi_detect/od/__init__.py
Comment (ascillitoe) [inline alibi_detect/od/__init__.py:5]: Assume this shouldn't be here?

Comment (jklaise) [reply alibi_detect/od/__init__.py:5]: +1

File: setup.cfg
Comment (ascillitoe) [inline setup.cfg:46]: What is the deal with `all` being commented out?

Comment (mauicv) [reply setup.cfg]: I'm not sure to be honest, its been uncommented now anyway

**Review by ascillitoe — APPROVED** [review]
LGTM! Assuming CI passes...

**Pull request merged**
